### PR TITLE
Changes to Prisma Default Value Handling

### PR DESCRIPTION
**Pull request by pimeys**
Moving this article here until it's finished and I can just dump it to Notion.

[PDF](https://github.com/pimeys/writing/blob/datetime-issue/prisma/Changes%20to%20Prisma%20Default%20Value%20Handling.pdf) [HTML](https://github.com/pimeys/writing/blob/datetime-issue/prisma/Changes%20to%20Prisma%20Default%20Value%20Handling.md)

### CODE REVIEW

File: prisma/Changes to Prisma Default Value Handling.org [hunk 1]
```diff
@@ -0,0 +1,887 @@
...
+
+In the next chapters we go through the problems one by one. In general we see Migration Engine failing with datetime defaults in older MySQL versions, Introspection Engine not really able to read datetime defaults in any database and Client will produce wrong values if the user uses a non-UTC timezone for every type except timestamps.
+
+** The Result in Different Datetime Types
+
+In the following examples we take a look what actually gets stored when we either use a non-UTC default value or the client uses a non-UTC value in their insert query. Failures in this chapters are marked with =C= in the tables.
+
+** MySQL
+
+In our example we store a default value of ="1975-08-19T23:15:30+07:00"= to different types of datetime columns, observing what goes wrong. In all cases the Query Engine converts the value to =1975-08-19T16:15:30Z= before storing (see below where we modify the value).
```
Comment (janpio) [inline prisma/Changes to Prisma Default Value Handling.org:66]: The first value is not rendered properly by GH. Should this maybe be:
```suggestion
In our example we store a default value of =1975-08-19T23:15:30+07:00= to different types of datetime columns, observing what goes wrong. In all cases the Query Engine converts the value to =1975-08-19T16:15:30Z= before storing (see below where we modify the value).
```

Comment (janpio) [reply prisma/Changes to Prisma Default Value Handling.org:66]: In the Notion version this string is also wrapped in `@default(...)`

Comment (pimeys) [reply prisma/Changes to Prisma Default Value Handling.org:66]: Yeah we don't really have to. I found it to have a bit too much of clutter. We have it in the schema examples, but having too much of code in a text block makes it harder to read.

File: prisma/Changes to Prisma Default Value Handling.org [hunk 2]
```diff
@@ -0,0 +1,887 @@
...
+
+*** @db.DateTime
+
+Storing the value =1975-08-19 16:15:30=. In the database we have values stored from other systems respecting the database default with the value of =1975-08-19 23:15:30=. We have two different default values in the database because of Prisma at this point.
+
+This is the default native type Prisma uses for datetimes.
+
+*** @db.Time
+
+Storing the value =16:15:30=. Other systems store the given =23:15:30= and now we have two different default values in the database with no idea which ones are correct.
```
Comment (janpio) [inline prisma/Changes to Prisma Default Value Handling.org:76]: What does "Storing the value" mean here? What are you doing exactly to get to this result? The steps should probably be explained under "The Result in Different Datetime Types" as "Test setup" or similar.

Comment (janpio) [reply prisma/Changes to Prisma Default Value Handling.org:76]: (This applies to all of these tests of course - it is unclear what "store" means in the context of Migrations to me. Part of the table definition?)

File: prisma/Changes to Prisma Default Value Handling.org [hunk 3]
```diff
@@ -0,0 +1,887 @@
...
+
+*** @db.Date
+
+In this example we have in the PSL the default =1975-08-19T00:05:30+01:00=. Due to the conversion to UTC, the value gets converted into =1975-08-18T23:05:30Z=. We store the value of =1975-08-18= and other systems writing to the database use the value =1975-08-19=.
```
Comment (janpio) [inline prisma/Changes to Prisma Default Value Handling.org:80]: Shouldn't we store a value like `1975-08-19T00:00:00+02:00` technically? (only Date part filled, time empty)

File: prisma/Changes to Prisma Default Value Handling.org [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+
+*** @db.Timestamp
+
+A timestamp value is always a number of (micro)seconds since =1970-01-01 00:00:00= in UTC. Converting between timezones does not change the underlying value. The =TIMESTAMP= type is the only one that is not set wrong by Prisma if the user is using a non-UTC timezone.
+
+** PostgreSQL
+
+In our example we store a default value of ="1975-08-19T23:15:30+07:00"= to different types of datetime columns, observing what goes wrong. In all cases the Query Engine converts the value to =1975-08-19T16:15:30Z= before storing.
```
Comment (janpio) [inline prisma/Changes to Prisma Default Value Handling.org:88]:
```suggestion
In our example we store a default value of =1975-08-19T23:15:30+07:00= to different types of datetime columns, observing what goes wrong. In all cases the Query Engine converts the value to =1975-08-19T16:15:30Z= before storing.
```

File: prisma/Changes to Prisma Default Value Handling.org [hunk 5]
```diff
@@ -0,0 +1,887 @@
...
+
+We have in the PSL the default =1975-08-19T00:05:30+01:00=. Due to the conversion to UTC, the value gets converted into =1975-08-18T23:05:30Z=. We store the value of =1975-08-18= and other systems writing to the database use the value =1975-08-19=.
+
+*** @db.Time
+
+Storing the value =16:15:30=. Other systems store the given =23:15:30= and now we have two different default values in the database.
+
+*** @db.DateTimeOffset
+
+We don't really strong a wrong value here or lose information as we do with the other types. We just store a wrong offset in for the =datetimeoffset= type. We store =1975-08-19T16:15:30Z= and any other instance that is not Prisma writes =1975-08-19T23:15:30+07:00=. This data can be corrected, but is still not correct.
```
Comment (janpio) [inline prisma/Changes to Prisma Default Value Handling.org:132]:
```suggestion
We don't really store a wrong value here or lose information as we do with the other types. We just store a wrong offset in for the =datetimeoffset= type. We store =1975-08-19T16:15:30Z= and any other instance that is not Prisma writes =1975-08-19T23:15:30+07:00=. This data can be corrected, but is still not correct.
```

File: prisma/Changes to Prisma Default Value Handling.org [hunk 6]
```diff
@@ -0,0 +1,887 @@
...
+| Date            | C I                    |
+| Time            | C I                    |
+
+In the next chapters we go through the problems one by one. In general we see Migration Engine failing with datetime defaults in older MySQL versions, Introspection Engine not really able to read datetime defaults in any database and Client will produce wrong values if the user uses a non-UTC timezone for every type except timestamps.
+
+** The Result in Different Datetime Types
+
+In the following examples we take a look what actually gets stored when we either use a non-UTC default value or the client uses a non-UTC value in their insert query. Failures in this chapters are marked with =C= in the tables.
+
+** MySQL
```
Comment (janpio) [inline prisma/Changes to Prisma Default Value Handling.org:64]: How can we even test this when like described under "Default Values in Migrations" a migration of such a schema does not even work on MySQL? Or did these all use 8.0.19 to these tests?

File: prisma/Changes to Prisma Default Value Handling.org [hunk 7]
```diff
@@ -0,0 +1,887 @@
...
+**** Date
+
+#+begin_src prisma
+model foo {
+  id Int      @id @default(autoincrement())
+  a  DateTime @default("1995-05-02T16:20:00+07:00") @db.Date
+}
+#+end_src
+
+Push works. Introspection returns:
```
Comment (janpio) [inline prisma/Changes to Prisma Default Value Handling.org:301]: Per documentation migrating with a timezone element should only work for `DATETIME` and `TIMESTAMP`, not `DATE` and `TIME`: https://dev.mysql.com/doc/refman/8.0/en/datetime.html#:~:text=In%20MySQL%208.0.19%20and%20later%2C%20you%20can%20specify%20a%20time%20zone%20offset%20when%20inserting%20a%20TIMESTAMP%20or%20DATETIME%20value%20into%20a%20table.%20See%20Section%C2%A09.1.3%2C%20%E2%80%9CDate%20and%20Time%20Literals%E2%80%9D%2C%20for%20more%20information%20and%20examples.

File: prisma/Changes to Prisma Default Value Handling.org (hunk 3 above)
Comment (pimeys) [reply prisma/Changes to Prisma Default Value Handling.org:88]: No. We convert to a Date. That takes out all parts that are not part of a Date.

Comment (janpio) [reply prisma/Changes to Prisma Default Value Handling.org:88]: Ah ok, so we just convert it back into a JS `Date` on the way back when reading, which then looks like a date with 00:00:00 (or similar, depending on DB).

File: prisma/Changes to Prisma Default Value Handling.org (hunk 6 above)
Comment (pimeys) [reply prisma/Changes to Prisma Default Value Handling.org:72]: We only test them with 8.0.23 or something.

File: prisma/Changes to Prisma Default Value Handling.org (hunk 7 above)
Comment (pimeys) [reply prisma/Changes to Prisma Default Value Handling.org:309]: You are right. I'll fix this.

**Pull request merged**